### PR TITLE
fix history size calc

### PR DESCRIPTION
**Pull request by natachadelarosa**

### CODE REVIEW

File: squid-blockexplorer/src/blocks/utils.ts [hunk 1]
```diff
@@ -99,7 +99,7 @@ export function createCall(
 export function calcSpacePledged(solutionRange: bigint): bigint {
   const MAX_U64 = 2n ** 64n - 1n;
   const SLOT_PROBABILITY = [1n, 6n];
-  const PIECE_SIZE = 31744n;
+  const PIECE_SIZE = 65536n;
```
Comment (jfrank-summit) [inline squid-blockexplorer/src/blocks/utils.ts:102]: Shouldn't `PIECE_SIZE` be the same in `calcSpacePledged` and `calcHistorySize`? Is there a reason not to set these outside of the functions and share the constants?

Comment (natachadelarosa) [reply squid-blockexplorer/src/blocks/utils.ts:102]: Is the same, one is just being calculated, will update shortly

**Review by jfrank-summit — COMMENTED** [review]
It looks like tests are failing.

File: squid-blockexplorer/src/blocks/utils.ts [hunk 2]
```diff
@@ -113,16 +114,25 @@ export function calcSpacePledged(solutionRange: bigint): bigint {
...
-  const PIECE_SIZE = 32 * 1024;
-  const PIECES_IN_SEGMENT = 256;
-  const RECORD_SIZE = PIECE_SIZE - (KZG_WITNESS_SIZE + KZG_COMMITMENT_SIZE);
-  const RECORDED_HISTORY_SEGMENT_SIZE = (RECORD_SIZE * PIECES_IN_SEGMENT) / 2;
-  const SEGMENT_SIZE =
-    (RECORDED_HISTORY_SEGMENT_SIZE / RECORD_SIZE) * PIECE_SIZE * 2;
-
-  return BigInt(segmentsCount * SEGMENT_SIZE);
+
+  const SLOT_PROBABILITY = BigInt(1) / BigInt(6);
```
Comment (nazar-pc) [inline squid-blockexplorer/src/blocks/utils.ts:118]: When working with integers (like here) `1/6` is always `0`, dividing by `0` in the formula below will always result in either NaN or Infinity or exception/error. See how division is done in monorepo codebase where we treat numerator and denominator as separate numbers for that exact reason.

**Review by nazar-pc — CHANGES_REQUESTED** [review]
Unfortunately this makes no sense to me at all.

`calcSpacePledged` judging by the name and description is the amount of space that farmers pledged to the network. It uses outdated formula from Gemini 1/2 and needs to be updated like I described in Slack. It wasn't.

`calcHistorySize` judging by the name and description is either size of archived or recorded history, not sure which one it should be, but old code was returning archived history size. The last formula there was correct for archived history before, just the size of the segment was wrong because piece size has changed. Also I'm not sure why it had to be so complicated before, it was just doing calculations in forward direction just to do calculations in reverse direction next instead of literally `PIECE_SIZE * PIECES_IN_SEGMENT * segmentCount` :man_shrugging:

The formula to which `calcHistorySize` was updated doesn't look make sense to me either, I have no idea what it calculates, looks like an arbitrary division and multiplication :confused:

**Comment by natachadelarosa** [comment]
> Unfortunately this makes no sense to me at all.
> 
> `calcSpacePledged` judging by the name and description is the amount of space that farmers pledged to the network. It uses outdated formula from Gemini 1/2 and needs to be updated like I described in Slack. It wasn't.
> 
> `calcHistorySize` judging by the name and description is either size of archived or recorded history, not sure which one it should be, but old code was returning archived history size. The last formula there was correct for archived history before, just the size of the segment was wrong because piece size has changed. Also I'm not sure why it had to be so complicated before, it was just doing calculations in forward direction just to do calculations in reverse direction next instead of literally `PIECE_SIZE * PIECES_IN_SEGMENT * segmentCount` 🤷‍♂️
> 
> The formula to which `calcHistorySize` was updated doesn't look make sense to me either, I have no idea what it calculates, looks like an arbitrary division and multiplication 😕

Looks like I didn't understand well last time, I thought the new `calc` was to calculate the archived history.

So the new `calc` should be implemented to get the space pledge to the network by the farmers.

And the archived history size just need to update the piece size.

Is it like this? @nazar-pc

**Comment by nazar-pc** [comment]
Well, my understanding was that it was about space pledged since that is a tricky part (and the [link](https://github.com/subspace/subspace/blob/f782f7297a6d61d8f22a3b10d201396fe30708fd/crates/subspace-runtime/src/lib.rs#L396-L399) I sent to the code was clearly in `TotalSpacePledged::get()` method).

At the same time history size is trivial: archived segments contain pieces, so just multiply 3 number and that is it. What was there before was taking a piece size, calculating record size from it, then calculating recorded history size in records just to then re-calculate it back into archived history in pieces. A bunch of math that essentially did nothing more than confusing the reader.

And it should be somewhat obvious that the only variable in the formula I provided was `solution_range`, while in the formula this PR has it is a constant set to 8 and variable is `segmentsCount`, something is clearly off.

I recommend you to take the formula right before the link as that is how you calculate things with integers without losing data and precision (like `1/6 == 0` for instance).

**Comment by natachadelarosa** [comment]
I think it should be okay now, @nazar-pc

File: squid-blockexplorer/src/blocks/utils.ts [hunk 3]
```diff
@@ -91,38 +91,42 @@ export function createCall(
...
+  const RECORD_CHUNKS = BigInt(32768);
+  const SOLUTION_RANGE = BigInt(8);
+  const SCALAR = BigInt(32);
+
+  const history_size =
+    (MAX_U64 * SLOT_PROBABILITY[0]) /
+    SLOT_PROBABILITY[1] /
+    (solutionRange * PIECE_SIZE) /
+    (RECORD_BUCKETS * RECORD_CHUNKS) /
+    (SOLUTION_RANGE * SCALAR);
```
Comment (nazar-pc) [inline squid-blockexplorer/src/blocks/utils.ts:115]: Why using `()`? It is not grouped in a way that would help understanding it and in fact it destroys the formula.

For example you were supposed to do `/ solutionRange * PIECE_SIZE` and you're doing essentially `/ solutionRange / PIECE_SIZE`, which is will result in a completely different value.

Please be very careful!

As mentioned before, take the actual code (not the high-level formula in the comment) and use that literally (of course you'll replace saturating multiplication with a regular one here).

**Review by nazar-pc — CHANGES_REQUESTED** [review]

**Review by nazar-pc — COMMENTED** [review]
Looks fine now formula-wise, but please use more descriptive names:
* `RECORD_BUCKETS` -> `RECORD_NUM_S_BUCKETS`
* `SCALAR` -> `SCALAR_FULL_BYTES`

And so on for others. And add link to monorepo's source code such that reviewer can make sense of it later and compare with latest revision if necessary.

**Review by nazar-pc — APPROVED** [review]

**Pull request merged**